core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java

```diff
@@ -0,0 +1,288 @@
+   */
+  private final BlockMasterClient mMasterClient;
+
+  /**
+   * The net address of the worker.
+   */
+  private final WorkerNetAddress mWorkerAddress;
+
+  /**
+   * The helper instance for sync related methods.
+   */
+  private final BlockMasterSyncHelper mBlockMasterSyncHelper;
+
+  /**
+   * The block worker responsible for interacting with Alluxio and UFS storage.
+   */
+  private final BlockWorker mBlockWorker;
+  /**
+   * Last System.currentTimeMillis() timestamp when a heartbeat successfully completed.
+   */
+  private long mLastSuccessfulHeartbeatMs = 0;
+
+  private final BlockHeartbeatReporter mBlockHeartbeatReporter;
+
+  /**
+   * Creates a new instance of {@link SpecificMasterBlockSync}.
+   *
+   * @param blockWorker the {@link BlockWorker} this syncer is updating to
+   * @param masterClient the block master client
+   * @param heartbeatReporter the heartbeat reporter
+   */
+  public SpecificMasterBlockSync(
+      BlockWorker blockWorker,
+      BlockMasterClient masterClient, BlockHeartbeatReporter heartbeatReporter)
+      throws IOException {
+    mBlockWorker = blockWorker;
+    mWorkerId = blockWorker.getWorkerId();
+    mWorkerAddress = blockWorker.getWorkerAddress();
+    mMasterClient = masterClient;
+    mAsyncBlockRemover = new AsyncBlockRemover(mBlockWorker);
+    mBlockMasterSyncHelper = new BlockMasterSyncHelper(mMasterClient);
+    mMasterAddress = masterClient.getRemoteSockAddress();
+    mBlockHeartbeatReporter = heartbeatReporter;
+  }
+
+  private void registerWithMaster() {
+    RetryPolicy retry = createEndlessRetry();
+    while (retry.attempt()) {
+      try {
+        LOG.info("Registering with master {}", mMasterAddress);
+        // The content in the report can be cleared because registration will
+        // report these block information anyways.
+        mBlockHeartbeatReporter.clear();
+        registerWithMasterInternal();
+        LOG.info("Finished registration with {}", mMasterAddress);
+        return;
+      } catch (Exception e) {
+        LOG.error("Failed to register with master {}, error {}, retry count {} Will retry...",
+            mMasterAddress, e, retry.getAttemptCount());
+        mWorkerState = WorkerMasterRegistrationState.NOT_REGISTERED;
+      }
+    }
+    // Should not reach here because the retry is indefinite
+    ProcessUtils.fatalError(LOG, new RuntimeException(),
+        "Failed to register with master %s", mMasterAddress);
+  }
+
+  protected void registerWithMasterInternal()
+      throws IOException, FailedToAcquireRegisterLeaseException {
+    // The target master is not necessarily the one that allocated the workerID
+    LOG.info("Notify the master {} about the workerID {}", mMasterAddress, mWorkerId);
+    mMasterClient.notifyWorkerId(mWorkerId.get(), mWorkerAddress);
+    // TODO(elega) If worker registration to all masters happens at the same time,
+    // this might cause OOM issue. Consider limiting the worker registration concurrency.
```

Comment:
OOM on master or worker? The worker reg concurrency is by default controlled in https://docs.alluxio.io/os/user/stable/en/administration/Scalability-Tuning.html#worker-registration

Reply:
worker oom 
because worker need to get the block location list from the metastore for each master when it registers with the masters. If these registration happen at the same time, these block location list info might cause worker OOM

Reply:
I updated the comment